**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант №4

Набор данных: salary.csv

Атрибуты:

### 1.Чтение файла (набора данных)

In [1]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd
df = pd.read_csv(r"C:\Users\ar121\OneDrive\Рабочий стол\SUAI Labs\Введение в анализ данных\salary.csv", sep = ';')

### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [2]:
# применить метод head
df.head(20)

work_year employment_type                 job_title      salary  \
0      2020.0              FT            Data SCIENTIST     70000.0   
1      2020.0              FT      Product Data Analyst     20000.0   
2      2020.0              FT              Data Analyst     72000.0   
3      2020.0              FT            Data Scientist  11000000.0   
4      2020.0              FT            Data Scientist     45000.0   
5      2020.0              FT            Data Scientist   3000000.0   
6      2020.0              FT            Data Scientist     35000.0   
7      2020.0              FT              Data Analyst     85000.0   
8      2020.0              FT              Data Analyst      8000.0   
9      2020.0              FT             Data Engineer   4450000.0   
10     2020.0              FT      Product Data Analyst    450000.0   
11     2020.0              FT             Data Engineer     42000.0   
12     2020.0              FT             Data Engineer    720000.0   
13     2020.0              FT  Machine Learning Manager    157000.0   
14     2020.0              FT            Data Scientist     60000.0   
15     2020.0              FT              Data Analyst     41000.0   
16     2020.0              FT             Data Engineer     65000.0   
17     2020.0              FT              Data Analyst     10000.0   
18     2020.0              FT            Data Scientist     45760.0   
19     2020.0              FT             Data Engineer    106000.0   

    salary_in_usd employee_residence company_location company_size  
0         79833.0                 DE               DE            L  
1         20000.0                 HN               HN            S  
2         72000.0                 US               US            L  
3         35735.0                 HU               HU            L  
4         51321.0                 FR               FR            S  
5         40481.0                 IN               IN            L  
6         39916.0                 FR               FR            M  
7         85000.0                 US               US            L  
8          8000.0                 PK               PK        Large  
9         41689.0                 JP               JP            S  
10         6072.0                 IN               IN            L  
11        47899.0                 GR               GR            L  
12        33511.0                 MX               MX            S  
13       117104.0                 CA               CA            L  
14        68428.0                 GR               US            L  
15        46759.0                 FR               FR            L  
16        74130.0                 AT               AT            L  
17        10000.0                 NG               NG            S  
18        45760.0                 PH               US            S  
19       106000.0                 US               US            L

2.2 Оценка данных с помощью метода info.

In [3]:
# выполнит метод info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 401 entries, 0 to 400
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   work_year           401 non-null    float64
 1   employment_type     401 non-null    object 
 2   job_title           401 non-null    object 
 3   salary              398 non-null    float64
 4   salary_in_usd       401 non-null    float64
 5   employee_residence  401 non-null    object 
 6   company_location    401 non-null    object 
 7   company_size        401 non-null    object 
dtypes: float64(3), object(5)
memory usage: 25.2+ KB


2.3 Оценка данных с помощью метода describe.

In [4]:
# оцените числовые столбцы с помощью describe
df.describe()

work_year        salary  salary_in_usd
count   401.000000  3.980000e+02     401.000000
mean   2021.528678  2.888336e+05  105895.017456
std       0.678086  1.677081e+06   58183.664171
min    2020.000000  4.000000e+03    2859.000000
25%    2021.000000  6.700000e+04   65013.000000
50%    2022.000000  1.091400e+05  100000.000000
75%    2022.000000  1.500000e+05  140000.000000
max    2022.000000  3.040000e+07  412000.000000

### Промежуточный вывод о данных
Данные содержат 401 запись о работниках IT-сферы с 8 признаками, включая год выплаты заработной платы, тип работы, должность, зарплата за год, Зарплата в долларах, страна проживания, страна главного офиса и среднее кол-во людей в компании. <br>
Типы данных: <br>
Числовые: <br>
•	salary (float64) — зарплата в исходной валюте, содержит числовые значения. <br>
•	salary_in_usd (float64) — зарплата, приведённая к долларам США. <br>
•	work_year — год выплаты заработной платы <br>
Категориальные (object, на деле содержат ограниченный набор значений): <br>
•	employment_type — тип занятости (FT, PT, CT, FL). <br>
•	job_title — название должности. <br>
•	employee_residence — страна проживания сотрудника. <br>
•	company_location — страна расположения компании. <br>
•	company_size — размер компании (S, M, L).


---

**Сделать выводы. Вы должны понимать, что означает тот или иной столбец, чтобы
 ответить на вопросы на защите.**


 ---


 2.4 Оценка названий столбцов

In [5]:
# Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.
print(df.columns)

Index(['work_year', 'employment_type', 'job_title', 'salary', 'salary_in_usd',
       'employee_residence', 'company_location', 'company_size'],
      dtype='object')


В названиях колонок отсутствуют пробелы и кирилица, таким образом переименование не ялвляется необходимым

### 3. Проверка пропусков

In [6]:
# Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением).
print(df.isna().sum())

work_year             0
employment_type       0
job_title             0
salary                3
salary_in_usd         0
employee_residence    0
company_location      0
company_size          0
dtype: int64


In [7]:
df["salary"] = df["salary"].fillna(df["salary"].median())
print(df.isna().sum())

work_year             0
employment_type       0
job_title             0
salary                0
salary_in_usd         0
employee_residence    0
company_location      0
company_size          0
dtype: int64



---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с пропусками, а также знать другие способы работы с пропусками, чтобы
 ответить на вопросы на защите.**


 ---

### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [8]:
print("Дубликатов:", df.duplicated().sum())

Дубликатов: 55


In [9]:
# удалите дубликаты, если они есть
df = df.drop_duplicates()
print("Дубликатов:", df.duplicated().sum())
df = df.drop_duplicates().reset_index()

Дубликатов: 0


#### Проверка неявных дубликатов

In [10]:
print(df['job_title'].unique())
print(df['company_size'].unique())
print(df['employment_type'].unique())
print(df['company_location'].unique())
print(df['employee_residence'].unique())

['Data SCIENTIST' 'Product Data Analyst' 'Data Analyst' 'Data Scientist'
 'Data Engineer' 'Machine Learning Manager' 'Data Analytics Engineer'
 'Data Science Engineer' 'Machine Learning Developer'
 'Data Analytics Manager' 'Head of Data Science'
 'Head of Machine Learning' 'NLP Engineer' 'Data Analytics Lead'
 'DataScientist' 'Data AnalyticsManager']
['L' 'S' 'M' 'Large']
['FT' 'PT' 'FL']
['DE' 'HN' 'US' 'HU' 'FR' 'IN' 'PK' 'JP' 'GR' 'MX' 'CA' 'AT' 'NG' 'GB'
 'ES' 'IT' 'LU' 'PL' 'NL' 'IQ' 'UA' 'IL' 'RU' 'MT' 'CL' 'IR' 'BR' 'VN'
 'TR' 'DZ' 'MY' 'AU' 'CH']
['DE' 'HN' 'US' 'HU' 'FR' 'IN' 'PK' 'JP' 'GR' 'MX' 'CA' 'AT' 'NG' 'PH'
 'GB' 'ES' 'IT' 'PL' 'BG' 'NL' 'IQ' 'UA' 'SG' 'RU' 'MT' 'CL' 'RO' 'IR'
 'VN' 'BR' 'HK' 'TR' 'RS' 'AR' 'DZ' 'AU' 'CH']


In [11]:
# удалите дубликаты, если они есть
df['job_title'] = df['job_title'].replace('Data SCIENTIST', 'Data Scientist')
df['job_title'] = df['job_title'].replace('Data AnalyticsManager', 'Data Analytics Manager')
df['job_title'] = df['job_title'].replace('DataScientist', 'Data Scientist')
df['company_size'] = df['company_size'].replace('Large', 'L')
print(df['job_title'].unique())
print(df['company_size'].unique())

['Data Scientist' 'Product Data Analyst' 'Data Analyst' 'Data Engineer'
 'Machine Learning Manager' 'Data Analytics Engineer'
 'Data Science Engineer' 'Machine Learning Developer'
 'Data Analytics Manager' 'Head of Data Science'
 'Head of Machine Learning' 'NLP Engineer' 'Data Analytics Lead']
['L' 'S' 'M']


---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с дубликатами, знать, что такое явные и неявные дубликаты и способы работы с ними, чтобы ответить на вопросы на защите.**


 ---

### 5. Провека типов данных

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 346 entries, 0 to 345
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   index               346 non-null    int64  
 1   work_year           346 non-null    float64
 2   employment_type     346 non-null    object 
 3   job_title           346 non-null    object 
 4   salary              346 non-null    float64
 5   salary_in_usd       346 non-null    float64
 6   employee_residence  346 non-null    object 
 7   company_location    346 non-null    object 
 8   company_size        346 non-null    object 
dtypes: float64(3), int64(1), object(5)
memory usage: 24.5+ KB


In [13]:
# Проверьте типы данных, при необходимости измените типы данных, чтобы они соответствовали действительности.
df["work_year"] = df["work_year"].astype('int64')
df["salary"] = df["salary"].astype('int64')
df["salary_in_usd"] = df["salary_in_usd"].astype('int64')
df["employee_residence"] = df["employee_residence"].astype('category')
df["company_location"] = df["company_location"].astype('category')
df["company_size"] = df["company_size"].astype('category')
df["employment_type"] = df["employment_type"].astype('category')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 346 entries, 0 to 345
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   index               346 non-null    int64   
 1   work_year           346 non-null    int64   
 2   employment_type     346 non-null    category
 3   job_title           346 non-null    object  
 4   salary              346 non-null    int64   
 5   salary_in_usd       346 non-null    int64   
 6   employee_residence  346 non-null    category
 7   company_location    346 non-null    category
 8   company_size        346 non-null    category
dtypes: category(4), int64(4), object(1)
memory usage: 17.9+ KB


### Изменения в типах данных <br>
В ходе работы с данными были внесены следующие изменения в типы столбцов: <br>
•	work_year был приведён к типу int64, так как это целочисленные значения (годы), и хранение их в числовом формате позволяет корректно выполнять сортировку и арифметические операции. <br>
•	salary и salary_in_usd преобразованы к типу int64, поскольку зарплата является числовым показателем и не должна храниться как число с плавающей точкой. <br>
•	employee_residence, company_location, company_size и employment_type переведены в тип category. Эти поля содержат ограниченное количество уникальных значений (страны проживания, местоположение компании, размер компании, тип занятости). Использование категориального типа данных снижает объём занимаемой памяти и ускоряет операции фильтрации и группировки.


---

**Обратите внимание, что во всех вариантах необходимо сделать приведение типов. Будьте готовы на защите аргументировать проверку типов (почему выполнены те или иные преобразования).**


 ---

### 6. Группировка данных

#### Задание 1

Группировка - “work_year” и количество компаний каждого размера

In [14]:
# выполните группировку согласно варианту
df.groupby(["work_year","company_size"])["salary"].mean()

C:\Users\ar121\AppData\Local\Temp\ipykernel_27396\2589477629.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby(["work_year","company_size"])["salary"].mean()


work_year  company_size
2020       L               744522.727273
           M                91230.000000
           S               412997.333333
2021       L               720716.654545
           M               282933.500000
           S               198316.666667
2022       L               504426.086957
           M               125346.166667
           S                47750.000000
Name: salary, dtype: float64

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

Промежуточный вывод по полученной группировке: <br>
1.	Доминирующие группы: <br>
•	В 2021 году крупнейшие компании (L) имели наибольшее количество сотрудников — 55.<br>
•	В 2022 году лидерами стали компании среднего размера (M) — 173.<br>
2.	Динамика по годам:<br>
•	В 2020 году преобладали крупные компании (L — 22), но также значимо представлены малые (S — 15).<br>
•	В 2021 году наблюдается рост по всем категориям, особенно у компаний среднего размера (M — 30).<br>
•	В 2022 году отмечен резкий скачок числа сотрудников в компаниях среднего размера (M — 173), при снижении доли крупных (L — 23) и малых (S — 4).<br>
3.	Паттерны по размеру компаний:<br>
•	Компании малого размера (S) наиболее заметны в 2020 году (15), но к 2022 году их число резко снизилось (4).<br>
•	Крупные компании (L) сохраняют стабильные позиции, но их численность не растёт, а в 2022 даже снижается.<br>
•	Основной рост приходится на компании среднего размера (M), что указывает на расширение этого сегмента.<br>


---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 2

Группировка - “employment_type” и количество компаний по каждой
локации “company_location”. Создать датафрейм. Переименовать столбец с
количеством в “сount”. Отсортировать по возрастанию столбца “count”. 

In [15]:
# выполните группировку согласно варианту
grouped = (df.groupby(["employment_type", "company_location"], observed=True).size().reset_index(name="count"))

grouped = grouped.sort_values("count", ascending=True)

print(grouped)

   employment_type company_location  count
3               FT               BR      1
2               FT               AU      1
6               FT               CL      1
5               FT               CH      1
14              FT               IL      1
12              FT               HN      1
13              FT               HU      1
31              FT               VN      1
24              FT               NL      1
27              FT               RU      1
22              FT               MY      1
20              FT               MT      1
19              FT               LU      1
18              FT               JP      1
17              FT               IR      1
16              FT               IQ      1
29              FT               UA      1
34              PT               ES      1
33              PT               DZ      1
36              PT               NL      1
35              PT               IT      1
32              PT               DE      1
26         

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

Промежуточный вывод по полученной группировке:<br>
1.	Доминирующая группа <br>
•	Абсолютное большинство наблюдений приходится на сотрудников с полной занятостью (FT) в США (206 записей).<br>
•	Это указывает на то, что рынок труда США представлен наиболее широко и сильно влияет на общую картину.<br>
2.	Страны с заметным присутствием<br>
•	Помимо США, значительные значения у Великобритании (32), Канады (15), Индии (13), Франции и Германии (по 11) и Испании и Греции (по 9).<br>
•	Эти страны формируют «второй эшелон» по количеству сотрудников.<br>
3.	Страны с единичными значениями<br>
•	Многие страны (Бразилия, Австралия, Япония, Россия, Украина и др.) представлены всего одной записью, что отражает либо редкость данных из этих регионов, либо ограниченность выборки.<br>
4.	Особенности по типу занятости<br>
•	Почти все записи относятся к полной занятости (FT).<br>
•	Частичная занятость (PT) встречается только в отдельных странах Европы и Северной Африки (Испания, Италия, Германия, Нидерланды, Алжир).<br>
•	Свободная занятость (FL) представлена только в США (2 записи).<br>


---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 3

Сводная таблица (pivot_table) - средняя зарплата по годам
(”work_year”). Отсортировать по убыванию зп. Округлить до трёх знаков после
запятой. Переименовать столбец “salary” в “зарплата”.

In [16]:
# выполните сводную таблицу согласно варианту
pivot = pd.pivot_table(df, values='salary', index='work_year')

pivot = pivot.rename(columns={'salary': 'зарплата'})

pivot = pivot.round(3)
pivot = pivot.sort_values(by='зарплата', ascending=False)

pivot = pivot.reset_index()
print(pivot)

   work_year    зарплата
0       2020  548347.857
1       2021  501913.796
2       2022  167179.269


**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной сводной таблицы?`**

Промежуточный вывод по полученной группировке: <br>
1.	Динамика по годам<br>
•	В 2020 году средняя зарплата была наибольшей — ≈548 тыс.<br>
•	В 2021 году показатель снизился, но остался высоким — ≈502 тыс.<br>
•	В 2022 году наблюдается резкое падение — до ≈167 тыс.<br>
2.	Общая тенденция<br>
•	С 2020 по 2022 год прослеживается устойчивая нисходящая динамика зарплат.<br>
•	Особенно заметен обвал в 2022 году, где среднее значение более чем втрое меньше, чем в 2020–2021 годах.<br>
3.	Возможная интерпретация<br>
•	Это может быть связано с изменением структуры выборки: в 2022 году в датасет попало больше сотрудников из стран с низкими зарплатами или из компаний меньшего размера.<br>
•	Также возможны экономические причины (например, влияние кризиса или изменений на рынке труда).<br>


---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 4

Сводная таблица (pivot_table) - средняя зарплата в usd по должностям
(”job_title”) - строки, и по годам - столбцы. Отсортировать по возрастанию названия
должности (”job_title”) .

In [17]:
# выполните сводную таблицу согласно варианту
pivot = pd.pivot_table(df, values='salary_in_usd', index='job_title', columns='work_year', aggfunc='mean')

pivot = pivot.round(3)

pivot = pivot.sort_index(ascending=True)

pivot.columns.name = 'Год'
print(pivot)

Год                               2020        2021        2022
job_title                                                     
Data Analyst                 45547.286   79505.412   97199.321
Data Analytics Engineer            NaN   79732.333   20000.000
Data Analytics Lead                NaN         NaN  405000.000
Data Analytics Manager             NaN  126666.667  127485.000
Data Engineer                88162.000   79434.871  125380.833
Data Science Engineer              NaN   83705.000   60000.000
Data Scientist               85970.524   70215.773  130367.984
Head of Data Science               NaN   97500.000  195937.500
Head of Machine Learning           NaN         NaN   79039.000
Machine Learning Developer         NaN  100000.000   78791.000
Machine Learning Manager    117104.000         NaN         NaN
NLP Engineer                       NaN         NaN   37236.000
Product Data Analyst         13036.000         NaN         NaN


**`Сделайте вывод - интерпретацию - что можно увидеть из результата выполнения данного кода?`**

Промежуточный вывод по полученной группировке: <br>
1.	Рост зарплат по ключевым профессиям<br>
•	Data Scientist: с ≈85,970 в 2020 до ≈129,104 в 2022 — устойчивый рост, особенно в 2022.<br>
•	Data Engineer: незначительное снижение в 2021 (≈79,435), но затем рост до ≈125,381 в 2022.<br>
•	Data Analyst: постепенное увеличение — с ≈45,547 в 2020 до ≈97,199 в 2022.<br>
2.	Редкие должности с выбросами<br>
•	Data Analytics Lead (405,000 в 2022) и Head of Data Science (195,938 в 2022) имеют экстремально высокие значения, что связано с редкими позициями в выборке и, возможно, единичными наблюдениями.<br>
•	Product Data Analyst встречается только в 2020 году (≈13,036).<br>
3.	Должности с падением зарплат<br>
•	Machine Learning Developer: с 100,000 (2021) до 78,791 (2022).<br>
•	Head of Machine Learning: единственное значение в 2022 году — 79,039 (относительно низко для уровня «Head»).<br>
•	NLP Engineer: в 2022 году всего ≈37,236 — также низкий показатель, возможно из-за малой выборки.<br>
4.	Общая тенденция<br>
•	Основные роли в области Data Science и Data Engineering демонстрируют устойчивый рост доходов.<br>
•	Менеджерские и редкие позиции дают сильные колебания, что связано с малым количеством данных.<br>
•	Таблица показывает усиление спроса и роста зарплат в 2022 году именно для базовых ролей (Data Scientist, Data Engineer, Data Analyst).<br>


---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

### Вывод


***`Сделать расширенный вывод. `***

Вывод должен описывать результаты исследования: какой набор данных вы анализировали (предметная область), какие методы предобработки применили, какие выводы из группировок и сводных таблиц можно сделать (они могут повторять выводы выше) и т.п. Объём вывода, как правило не более 0.5-0.75 страницы А4, всё таки вывод должен быть конкретным (без теории и “воды”).


В ходе лабораторной работы был проанализирован набор данных salary.csv, содержащий информацию о заработных платах сотрудников в различных странах и компаниях. Данные включали: год выплаты заработной платы, тип занятости (PT — частичная, FT — полная, FL — фриланс), должность, зарплату в исходной валюте и в долларах США, страну проживания сотрудника, страну расположения главного офиса компании, а также размер компании (S — малая, M — средняя, L — крупная). <br>
Предобработка данных включала:<br>
•	проверку и корректировку названий столбцов;<br>
•	выявление и заполнение пропусков: числовые значения заменены медианами, категориальные признаки — модальными значениями;<br>
•	удаление дубликатов;<br>
•	приведение типов данных: годы и зарплаты переведены в int64, категориальные признаки (тип занятости, страна проживания, локация компании, размер компании) приведены к типу category.<br>
Ключевые выводы из анализа:<br>
1.	Динамика зарплат по годам: средние значения зарплат в долларах снизились с ≈548 тыс. в 2020 году до ≈502 тыс. в 2021, а затем резко упали в 2022 году до ≈167 тыс. Такая тенденция указывает на сильное падение доходов в выборке 2022 года, вероятно, связанное с изменением структуры данных или рыночными факторами.<br>
2.	Размер компаний и годы: в 2020–2021 годах лидировали крупные компании (L), однако в 2022 году на первый план вышли компании среднего размера (M — 173 сотрудника). При этом число сотрудников малых компаний (S) резко снизилось, что отражает возможную консолидацию рынка.<br>
3.	Группировка по типу занятости и локации компаний: абсолютное доминирование принадлежит полной занятости (FT), особенно в США (206 сотрудников). Другие страны представлены значительно меньше, зачастую единичными наблюдениями. Частичная занятость (PT) встречается преимущественно в Европе, а фриланс (FL) — только в США.<br>
4.	Сводная таблица по должностям и годам: профессии Data Scientist, Data Engineer и Data Analyst демонстрируют устойчивый рост зарплат к 2022 году, что отражает высокий спрос на специалистов в этих областях. В то же время редкие должности (например, Data Analytics Lead, Head of Data Science) показали очень высокие или нестабильные значения, что связано с малым числом наблюдений.
Таким образом, проведённая предобработка позволила очистить и структурировать данные, а анализ показал ключевые тенденции: доминирование полной занятости и рынка США, рост роли компаний среднего размера в 2022 году, а также устойчивый рост зарплат в ключевых профессиях в сфере Data Science при нестабильности редких должностей.<br>


### Дополнительное задание

**`Подробная формулировка задания`**
Создать столбец “Категория зарплаты в долларах” (с помощью
категоризации). Выделить минимум 3 категории (низкая, высокая, средняя),
фильтрацию для уровня зарплаты выбрать самостоятельно, аргументировать
выбор. Создать сводную таблицу: средняя и медианная зп в долларах по
категории зарплаты в долларах и среднему количество людей в компании. 

In [25]:
# код выполнения задания
q1 = df["salary_in_usd"].quantile(0.25)
q2 = df["salary_in_usd"].quantile(0.5)
q3 = df["salary_in_usd"].quantile(0.75)

def categorize_salary(salary):
    if salary <= q1:
        return "Низкая"
    elif salary <= q3:
        return "Средняя"
    else:
        return "Высокая"

df["Категория зарплаты в долларах"] = df["salary_in_usd"].apply(categorize_salary)
pivot = pd.pivot_table(
    df,
    values="salary_in_usd",
    index=["Категория зарплаты в долларах", "company_size"],
    aggfunc=["mean", "median"]
)

# Округлим для удобства
pivot = pivot.round(2)
print("\nСводная таблица:")
print(pivot)
print("Квартильные границы:")
print("Q1 =", q1, "| Q2 (медиана) =", q2, "| Q3 =", q3)


Сводная таблица:
                                                    mean        median
                                           salary_in_usd salary_in_usd
Категория зарплаты в долларах company_size                            
Высокая                       L                195675.00      172500.0
                              M                175190.92      165310.0
Низкая                        L                 37966.00       37825.0
                              M                 39120.45       43966.0
                              S                 29041.40       23600.5
Средняя                       L                 96938.16       96282.0
                              M                 98996.17      100000.0
                              S                 85984.71       87738.0
Квартильные границы:
Q1 = 60000.0 | Q2 (медиана) = 99025.0 | Q3 = 135000.0


C:\Users\ar121\AppData\Local\Temp\ipykernel_27396\55164804.py:15: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot = pd.pivot_table(
C:\Users\ar121\AppData\Local\Temp\ipykernel_27396\55164804.py:15: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot = pd.pivot_table(


Аргументация выбора значений в колонке «Категория зарплаты в долларах» 

Для выделения категорий заработных плат был выбран метод разбиения по квартилям:

Низкая зарплата — значения, меньшие или равные первому квартилю (Q1). Это 25% сотрудников с самыми низкими доходами.

Средняя зарплата — значения между первым и третьим квартилем (Q1–Q3). Это основной диапазон зарплат, который охватывает 50% сотрудников.

Высокая зарплата — значения, превышающие третий квартиль (Q3). Это 25% сотрудников с наибольшими доходами.

Такой способ категоризации выбран потому что:

Адаптивность к данным: границы не фиксированы заранее, а зависят от реального распределения зарплат в датасете.

Сбалансированность классов: каждая категория содержит сопоставимое количество наблюдений, что позволяет избежать перекоса, когда большинство сотрудников оказывается в одной категории.

Интерпретируемость: разделение по квартилям — стандартный статистический метод, который часто применяется в анализе доходов и социально-экономических данных.

Таким образом, использование квартилей позволяет корректно и объективно разделить сотрудников на группы по уровню зарплаты без субъективного выбора порогов.

***`Подробный вывод по заданию, описание полученных результатов`***